centreon/src/Core/ResourceAccess/Application/UseCase/AddRule/AddRule.php

```diff
@@ -0,0 +1,429 @@
+            $this->saveDatasetFiltersHierarchy($ruleId, $datasetId, $datasetFilter);
+
+            $this->addDatasetRelations(
+                $datasetId,
+                DatasetFilterTypeConverter::fromString($applicableFilter->getType()),
+                $applicableFilter->getResourceIds()
+            );
+
+            $index++;
+        }
+    }
+
+    /**
+     * @param NewRule $rule
+     *
+     * @return int
+     */
+    private function addRule(NewRule $rule): int
+    {
+        $this->info('Adding new rule');
+
+        return $this->writeRepository->add($rule);
+    }
+
+    /**
+     * @param int $ruleId
+     * @param NewRule $rule
+     */
+    private function linkContacts(int $ruleId, NewRule $rule): void
+    {
+        $this->info(
+            'AddRule: Linking contacts to the resource access rule',
+            ['ruleId' => $ruleId, 'contact_ids' => $rule->getLinkedContactIds()]
+        );
+
+        $this->writeRepository->linkContactsToRule($ruleId, $rule->getLinkedContactIds());
+    }
+
+    /**
+     * @param int $ruleId
+     * @param NewRule $rule
+     */
+    private function linkContactGroups(int $ruleId, NewRule $rule): void
+    {
+        $this->info(
+            'AddRule: Linking contact groups to the resource access rule',
+            ['ruleId' => $ruleId, 'contact_group_ids' => $rule->getLinkedContactGroupIds()]
+        );
+
+        $this->writeRepository->linkContactGroupsToRule($ruleId, $rule->getLinkedContactGroupIds());
+    }
+
+    /**
+     * This method will ensure that the ids provided for the dataset filters are valid
+     * and that the structure of the dataset is correct regarding the hierarchy defined in the DatasetFilter entity.
+     *
+     * @param AddRuleRequest $request
+     *
+     * @throws RuleException
+     * @throws \InvalidArgumentException
+     *
+     * @return DatasetFilter[]
+     */
+    private function validateAndCreateDatasetFiltersFromRequest(AddRuleRequest $request): array
+    {
+        $datasetFilters = [];
+
+        $validateAndBuildDatasetFilter = function (
+            array $data,
+            ?DatasetFilter $parentDatasetFilter
+        ) use (&$validateAndBuildDatasetFilter, &$datasetFilter): void {
+            // In any case we want to make sure that resources provided are valid
+            // Then at DatasetFilter creation the hierarchy is validated
+            $this->validator->assertIdsAreValid($data['type'], $data['resources']);
+
+            // first iteration we want to create the root filter
+            if ($datasetFilter === null) {
+                $datasetFilter = new DatasetFilter(
+                    type: $data['type'],
+                    resourceIds: $data['resources']
+                );
+                if ($data['dataset_filter'] !== null) {
+                    $validateAndBuildDatasetFilter($data['dataset_filter'], null);
+                }
+            } else {
```

Comment:
I suggest breaking the flow to reduce the cognitive complexity.
> early-return like
```
                }
                return;
            }
```

Reply:
That will be done later on